pkg/tsdb/elasticsearch/client/index_pattern.go

```diff
@@ -247,72 +247,101 @@ var datePatternReplacements = map[string]string{
-		base = parts[0]
-		if len(parts) == 2 {
-			datePattern = parts[1]
+	baseStart := strings.Index(pattern, "[")
+	for baseStart != -1 {
+		var datePattern string
+
+		baseEnd := strings.Index(pattern, "]")
+		base = pattern[baseStart+1 : baseEnd]
+		bases = append(bases, base)
+		if baseStart == 0 {
+			isBaseFirst = true
 		} else {
-			datePattern = base
-			base = ""
+			datePattern = pattern[:baseStart]
+			formatted := t.Format(patternToLayout(datePattern))
+			formattedDatePatterns = append(formattedDatePatterns, formatted)
+		}
+
+		if len(pattern) <= baseEnd+1 {
+			break
 		}
-		ltr = true
-	} else if strings.HasSuffix(pattern, "]") {
-		parts := strings.Split(strings.TrimRight(pattern, "]"), "[")
-		datePattern = parts[0]
-		if len(parts) == 2 {
-			base = parts[1]
+		pattern = pattern[baseEnd+1:]
+
+		baseStart = strings.Index(pattern, "[")
+		if baseStart == -1 {
+			datePattern = pattern
 		} else {
-			base = ""
+			datePattern = pattern[:baseStart]
+			pattern = pattern[baseStart:]
+			baseStart = 0
 		}
-		ltr = false
+		formatted := t.Format(patternToLayout(datePattern))
+		formattedDatePatterns = append(formattedDatePatterns, formatted)
 	}
 
-	formatted := t.Format(patternToLayout(datePattern))
-
-	if strings.Contains(formatted, "<std") {
-		isoYear, isoWeek := t.ISOWeek()
-		isoYearShort := fmt.Sprintf("%d", isoYear)[2:4]
-		formatted = strings.ReplaceAll(formatted, "<stdIsoYear>", fmt.Sprintf("%d", isoYear))
-		formatted = strings.ReplaceAll(formatted, "<stdIsoYearShort>", isoYearShort)
-		formatted = strings.ReplaceAll(formatted, "<stdWeekOfYear>", fmt.Sprintf("%02d", isoWeek))
-
-		formatted = strings.ReplaceAll(formatted, "<stdUnix>", fmt.Sprintf("%d", t.Unix()))
-
-		day := t.Weekday()
-		dayOfWeekIso := int(day)
-		if day == time.Sunday {
-			dayOfWeekIso = 7
+	for i, formatted := range formattedDatePatterns {
+		if strings.Contains(formatted, "<std") {
+			isoYear, isoWeek := t.ISOWeek()
+			isoYearShort := fmt.Sprintf("%d", isoYear)[2:4]
+			formatted = strings.ReplaceAll(formatted, "<stdIsoYear>", fmt.Sprintf("%d", isoYear))
+			formatted = strings.ReplaceAll(formatted, "<stdIsoYearShort>", isoYearShort)
+			formatted = strings.ReplaceAll(formatted, "<stdWeekOfYear>", fmt.Sprintf("%02d", isoWeek))
+
+			formatted = strings.ReplaceAll(formatted, "<stdUnix>", fmt.Sprintf("%d", t.Unix()))
+
+			day := t.Weekday()
+			dayOfWeekIso := int(day)
+			if day == time.Sunday {
+				dayOfWeekIso = 7
+			}
+
+			formatted = strings.ReplaceAll(formatted, "<stdDayOfWeek>", fmt.Sprintf("%d", day))
+			formatted = strings.ReplaceAll(formatted, "<stdDayOfWeekISO>", fmt.Sprintf("%d", dayOfWeekIso))
+			formatted = strings.ReplaceAll(formatted, "<stdDayOfYear>", fmt.Sprintf("%d", t.YearDay()))
+
+			quarter := 4
+
+			switch t.Month() {
+			case time.January, time.February, time.March:
+				quarter = 1
+			case time.April, time.May, time.June:
+				quarter = 2
+			case time.July, time.August, time.September:
+				quarter = 3
+			}
```

Comment:
Can you move this outside of the loop? It would help in keeping the code as readable as possible.

Reply:
Thanks for the review - refactored all of your suggested changes into 878a1bc